src/App.tsx

```diff
@@ -40,39 +38,87 @@ class App extends Component<object, AppState> {
         return response.json();
       })
       .then((data: Spell[]) => {
-        const results = searchTerm
+        const filtered = searchTerm
           ? data.filter((spell) =>
               spell.name.toLowerCase().includes(searchTerm.toLowerCase())
             )
           : data;
-        this.setState({ results, loading: false });
+        setResults(filtered);
+        setLoading(false);
       })
       .catch((error) => {
-        this.setState({ error: error.message || undefined, loading: false });
+        setError(error.message || undefined);
+        setLoading(false);
       });
   };
 
-  throwError = () => {
-    this.setState(() => {
-      throw new Error('Test error');
-    });
+  useEffect(() => {
+    handleSearch('');
+  }, []);
+
+  // Pagination logic
+  const totalPages = Math.ceil(results.length / ITEMS_PER_PAGE);
+  const paginatedResults = results.slice(
+    (page - 1) * ITEMS_PER_PAGE,
+    page * ITEMS_PER_PAGE
+  );
+
+  const handlePageChange = (newPage: number) => {
+    setSearchParams({ page: String(newPage) });
   };
 
-  render() {
-    const { results, error, loading } = this.state;
+  const details = searchParams.get('details');
 
-    return (
-      <div>
-        <Header onSearch={this.handleSearch} />
-        <Results
-          results={results}
-          error={error}
-          loading={loading}
-          throwError={this.throwError}
-        />
+  const handleShowDetails = (name: string) => {
+    setSearchParams({ page: String(page), details: name });
+  };
+
+  const handleCloseDetails = () => {
+    setSearchParams({ page: String(page) });
+  };
+
+  return (
+    <div>
+      <nav style={{ marginBottom: 16 }}>
+        <Link to="/">Home</Link> | <Link to="/about">About</Link>
+      </nav>
+      <div style={{ display: 'flex' }}>
+        <Routes>
+          <Route
+            path="/"
+            element={
+              <div style={{ display: 'flex', width: '100%' }}>
+                <div style={{ flex: 1 }}>
+                  <Header onSearch={handleSearch} />
+                  <Results
+                    results={paginatedResults}
+                    error={error}
+                    loading={loading}
+                    onCardClick={handleShowDetails}
+                  />
+                  {!loading && (
+                    <Pagination
+                      currentPage={page}
+                      totalPages={totalPages}
+                      onPageChange={handlePageChange}
+                    />
+                  )}
+                </div>
+                {details && (
```

Comment:
it will be better

```js
                const currentSpell = results.find((spell) => spell.name === details);

               //

                {currentSpell && (
                  <SpellDetails
                    spell={currentSpell}
                    onClose={handleCloseDetails}
                  />
                )}
```